Update Dockerfile

dschaper opened the pull request:
This is a lot better than before, added back SSL capability, and user can be set to UID 999 as before. (Just needed to delgroup the PING group at 999, but that shouldn't be a problem.) Some refactoring and moving some ENV's to ARGS that can be overridden on build via comandline now. Set up for an docker-compose build with `environments:` to customize the actual build.

dschaper committed:
Update Dockerfile

This is a lot better than before, added back SSL capability, and user can be set to UID 999 as before. (Just needed to delgroup the PING group at 999, but that shouldn't be a problem.) Some refactoring and moving some ENV's to ARGS that can be overridden on build via comandline now. Set up for an docker-compose build with `environments:` to customize the actual build.

dschaper committed:
Use official base

Split this out as separate commit so that it can be left out and the other commit can be cherry picked if that's determined to be the best, but it works just fine on the `openjdk` alpine base.

dschaper committed:
Curl/BSDTAR

Curl for ssl compatability and following any Jetbrains site changes (follows 302's). Use BSDTAR from libarchive-tools to unzip the delivery file and set permissions directly on the files as they are unzipped.

esycat commented:
> added back SSL capability

Could you please clarify what do you mean by that? I am probably missing something.

> and user can be set to UID 999 as before

I'm not sure that it is worth returning the old UID. It was `999` only because that is what Ubuntu assigned by default to a new user, and I did not pay attention to that as coincidently `docker` user on my hosts also had UID `999`. It seems to be more reliable to hardcode a specific value instead. I see that some of the “official” images do exactly that.

dschaper commented:
The ssl was adding back the ability to pull from Jetbrains via https, with the straight busybox solution you would only be able to pull via http, as the busybox `wget` doesn't do well with security. So that was a commit I closed out and replaced with `curl` that will also follow redirects and do some other fun things.

I just used the old UID/GID as it was defined in the Ubuntu images and continuing on would prevent users having to change local volume permissions on upgrade. Not anything that can't be overridden or changed. I squashed my commits, otherwise it would be easier to cherry-pick out the stuff you like and leave the stuff you don't. (Easier, just change the Dockerfile and close out this PR and no merge, either way is good, I'm not looking for the PR points :) 

I'm still trying to figure out the best way to handle UID mapping, looking over userspace documentation now to see if there's a new way post docker 1.10 but nothing pops out yet.

dschaper commented:
My commits tend to be fast and furious with lots of branching, so they get really verbose unless I squash a few.

dschaper committed:
Big change

A lot of changes, User now has APP_DIR as his homedir, Refactored and squashed a lot of silly commits.

dschaper committed:
Update Dockerfile
